Add functions for non-ptr style virtual calls in GDExtension

fuzzybinary opened the pull request:
This adds two functions to `GDExtensionClassCreationInfo2` that allow for developers to supply a generic virtual call function along with user data to be sent to that call.

If `get_virutal_call_data_func` is not null, extensions call this function to get user data to pass to a supplied `call_virtual_func`. Both must be provided is one is provided.

If `get_virtual_call_data_func` is null, Godot falls back to the old `get_virtual_func` logic.

Fixes #63275

fuzzybinary committed:
Add functions for non-ptr style virtual calls in GDExtension

This adds two functions to `GDExtensionClassCreationInfo` that allow for developers to supply a generic virtual call function along with user data to be sent to that call.

If `get_virutal_call_data_func` is not null, extensions call this function to get user data to pass to a supplied `call_virtual_with_data_func`. Both must be provided is one is provided.

If `get_virtual_call_data_func` is null, Godot falls back to the old `get_virtual_func` logic.

Fixes #63275

Co-authored-by: David Snopek <dsnopek@gmail.com>

dsnopek commented on core/object/make_virtuals.py:
```diff
@@ -16,14 +16,24 @@
     if (unlikely(_get_extension() && !_gdvirtual_##m_name##_initialized)) {\\
-        /* TODO: C-style cast because GDExtensionStringNamePtr's const qualifier is broken (see https://github.com/godotengine/godot/pull/67751) */\\
-        _gdvirtual_##m_name = (_get_extension() && _get_extension()->get_virtual) ? _get_extension()->get_virtual(_get_extension()->class_userdata, (GDExtensionStringNamePtr)&_gdvirtual_##m_name##_sn) : (GDExtensionClassCallVirtual) nullptr;\\
+        _gdvirtual_##m_name = (GDExtensionClassCallVirtual) nullptr;\\
```
We probably shouldn't keep `GDExtensionClassCallVirtual` as the type on `_gdvirtual_##m_name`, because that could be confusing. We should either change this to `void *` or a "void function pointer" like `GDExtensionInterfaceFunctionPtr` (to ensure the correct alignment on platforms where that might matter). We can then cast to `GDExtensionClassCallVirtual` before calling it.

fuzzybinary replied:
Changed to a void* and also renamed the variable to `_gdvirtual_##m_name##_data`

dsnopek replied:
Thanks! I don't know that the change to `_gdvirtual_##m_name##_data` is really necessary, but it's fine :-)

fuzzybinary replied:
Wanted to be *really* clear what it was :-)

dsnopek reviewed (commented):
Thanks! Overall, I personally think this is a good approach.

This is something that will definitely need a bunch of discussion, though, perhaps at a meeting. It'd also be really great if we could get folks working on different language bindings (particularly those that were having trouble with the old way) to see if it would work for them.

fuzzybinary commented on core/extension/gdextension_interface.h:
```diff
@@ -283,6 +285,12 @@ typedef struct {
 	GDExtensionClassGetVirtual get_virtual_func; // Queries a virtual function by name and returns a callback to invoke the requested virtual function.
+	// Can be used instead of `get_virtual_func`. Returns user data (non-null) that will be passed back when calling `call_virtual_func`.
+	// Returning `null` from this function signals to Godot that the virtual function is not overridden.
+	// Data provided to this function must live until the GDExtensionInstance is freed, but should be deleted when `free_instance_func` is called.
```
Adjusted the language a bit. Let me know what you think.

dsnopek replied:
Hm. Thinking about it some more, I'm not sure I still like my suggestion either :-)

The GDExtension really doesn't know when the `GDExtensionInterface` will be freed. I think all it can do is keep the data alive until the GDExtension is deinitialized. I'll put that in a new suggestion in a moment...

dsnopek commented on core/extension/gdextension_interface.h:
```diff
@@ -283,6 +285,12 @@ typedef struct {
 	GDExtensionClassCreateInstance create_instance_func; // (Default) constructor; mandatory. If the class is not instantiable, consider making it virtual or abstract.
 	GDExtensionClassFreeInstance free_instance_func; // Destructor; mandatory.
 	GDExtensionClassGetVirtual get_virtual_func; // Queries a virtual function by name and returns a callback to invoke the requested virtual function.
+	// Can be used instead of `get_virtual_func`. Returns user data (non-null) that will be passed back when calling `call_virtual_func`.
+	// Returning `null` from this function signals to Godot that the virtual function is not overridden.
+	// Data provided to this function should be managed by the extension and must live until the GDExtensionInstance.
```
What about:

```suggestion
	// Can be used instead of `get_virtual_func`. Returns user data that will be passed to `call_virtual_func`.
	// Returning `NULL` from this function signals to Godot that the virtual function is not overridden.
	// Data returned from this function should be managed by the extension and must be valid until the extension is deinitialized.
```

I got a little carried away and made a whole bunch of changes that I think simplify the language a bit :-)

But the truly important change is end of the last sentence: "and must be valid until the extension is deinitialized"

dsnopek commented:
Thanks, this is looking good!

There's some CI failures from build errors in unrelated code. Rebasing on the latest `master` may fix that?

AThousandShips commented on core/object/make_virtuals.py:
```diff
@@ -16,15 +16,26 @@
-		_gdvirtual_##m_name(_get_extension_instance(),$CALLPTRARGPASS,$CALLPTRRETPASS);\\
-		$CALLPTRRET\\
+        if (_get_extension()->get_virtual_call_data) {\\
+            if (_get_extension()->call_virtual_with_data) {\\
```
In this path the method may never be called yet the method returns true, meaning the return value is uninitialised

fuzzybinary replied:
👍 Good catch. I'll have to think about how to fix without triggering clang linters.

AThousandShips replied:
Probably good indication to go over the code to ensure it does what it is supposed to do, good that the CI catches unassigned data related things, though it's very vague in the errors

fuzzybinary commented:
This should now be ready. I've moved the new members into the GDExtensionClassCreationInfo2 struct.

@AThousandShips I think I fixed the uninitialized ptr return variable. Can you take another look and see if it makes sense to you?

AThousandShips commented:
Will take a look!

dsnopek commented on core/object/make_virtuals.py:
```diff
@@ -2,7 +2,7 @@
 StringName _gdvirtual_##m_name##_sn = #m_name;\\
 mutable bool _gdvirtual_##m_name##_initialized = false;\\
-mutable GDExtensionClassCallVirtual _gdvirtual_##m_name = nullptr;\\
+mutable void* _gdvirtual_##m_name##_data = nullptr;\\
```
Personally, I think I'd still prefer if this kept the old variable name. Since this could be either a function pointer or user data, I don't think adding `_data` to the variable name really helps with clarity - if anything it's slightly more confusing since this might _not_ be data.

fuzzybinary replied:
I'll change it back. I was actually thinking the opposite, that using `_data` would make it more obvious that it was doing multiple things (serving as a function pointer or serving as callback data). But I'm not super attached to a name change.

dsnopek reviewed (commented):
This is looking great to me! I only have the one naming related note.

I think this should get some more feedback from other GDExtension team members, though, perhaps discussing it at the next meeting.

Also, the CI is failing on building godot-cpp, but I don't think that we can fix that. :-/ It's a circular dependency: godot-cpp will need to be updated for this change, but godot-cpp depends on the Godot change.

I think this will come up again as we make further changes to the new structs before Godot 4.2. I'm not sure what we can do about that? We could maybe have the Godot CI use the `4.1` branch of godot-cpp instead of `master`?

AThousandShips commented:
I think the code paths look good now and the CI errors have gone away

dsnopek commented:
PR https://github.com/godotengine/godot/pull/81238 should fix the CI. If/when it's merged, this one will just need a rebase.

fuzzybinary commented:
Rebased -- not sure if the CI error is something on my end?

AThousandShips commented:
It's random and unrelated to this PR

dsnopek reviewed (commented):
Discussed at the GDExtension meeting, and we think this should be evaluated in the context of more language bindings before committing to this interface. It's possible this is more complex than needed and only passing the method name is required, although, it's also possible that this interface is needed to provide performance optimizations for certain languages on par with what Godot can already do for C++.

dsnopek commented:
I've been thinking about this a bit after the discussion at the GDExtension meeting: I think it's possible that after enough language bindings have been created for GDExtension that the approach here could turn out to be overkill; but I don't think it could turn out to be "underkill" :-)

For language bindings that really only need the method name to call a virtual method, then `get_virtual_call_data()` is just an existence check for the virtual method with a kind of weird API. For those use cases, a `has_virtual_func()` and `call_virtual_func()` would have been sufficient and would certainly be a friendlier API. But, developers can still do what they need to do with the less friendly API in this PR.

However, if a language binding wants to improve virtual method call performance by caching a pointer to the function data (which is something that language bindings like godot-cpp and Rust already can do, because they can easily generate function pointers to C functions and use the existing `get_virtual_func()`), it wouldn't be possible with the simpler, friendlier API; that would require something like the API proposed in this PR.

This is all a long winded way to say: maybe we should just go with the approach in the PR. If we eventually discover that, yes, this was overkill, we can simplify it in Godot 5 :-)

fuzzybinary commented:
For what it's worth, the more complicated method you're describing is actually what I'm doing in Dart. See [here](https://github.com/fuzzybinary/godot_dart/blob/845db37c22653f24d88a0524c1da08b64df62246/src/cpp/dart_bindings.cpp#L653). Dart gives me C-style function pointer for a given function on the class, but because I needed to wrap calling it in some extra code, I couldn't use the original method.

So even if this is more complicated it is for sure giving me higher performance than if we were just sending in the name.

fuzzybinary commented:
Renamed variables and rebased onto master.

dsnopek commented:
@fuzzybinary:

> So even if this is more complicated it is for sure giving me higher performance than if we were just sending in the name.

Ah, thanks for sharing how you're using this in your Dart bindings! If you're able to take advantage of the API in order to cache pointers to the underlying Dart functions, I think that's a solid justification for this approach. It allows calls to virtual functions defined in Dart to avoid a lookup on every call, just like with the C++ binding. (Personally, I don't want any language bindings forced to be "second class citizens" just because they can't easily generate new functions with C linkage)

> Renamed variables and rebased onto master.

Thanks! The suggestion from the meeting was to also rename the `p_userdata` argument to `p_class_userdata` on the other related function pointer typedefs as well, for consistency. So, also on `GDExtensionClassCreateInstance` and `GDExtensionClassFreeInstance` too.

fuzzybinary commented:
👍 I can do that, I was thinking it would be better to keep these changes localized to virtuals but I can change it in this PR if we want.

dsnopek reviewed (approved):
Thanks! This is looking great to me :-)

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -304,6 +306,12 @@ typedef struct {
 	GDExtensionClassCreateInstance create_instance_func; // (Default) constructor; mandatory. If the class is not instantiable, consider making it virtual or abstract.
 	GDExtensionClassFreeInstance free_instance_func; // Destructor; mandatory.
 	GDExtensionClassGetVirtual get_virtual_func; // Queries a virtual function by name and returns a callback to invoke the requested virtual function.
+	// Can be used instead of `get_virtual_func`. Returns user data that will be passed to `call_virtual_func`.
+	// Returning `NULL` from this function signals to Godot that the virtual function is not overridden.
+	// Data returned from this function should be managed by the extension and must be valid until the extension is deinitialized.
+	GDExtensionClassGetVirtuaCallData get_virtual_call_data_func;
+	// Used to call virtual functions when `get_virtual_call_data_func` is not null.
+	GDExtensionClassCallVirtualWithData call_virtual_func;
```
What is missing is a brief guide regarding when to use the new approach (`get_virtual_call_data_func` + `call_virtual_func`) vs. when to use the old one (`get_virtual_func`). In particular, it should be mentioned that they both coexist with separate use cases, one doesn't supersede the other.

Also, maybe mention that the data returned by this function is passed back to `call_virtual_function`.

Last, typo `GDExtensionClassGetVirtuaCallData` -> "Virtua"

fuzzybinary replied:
Updated docs, let me know if that's clearer.

Bromeon replied:
This text is still the same?
```cpp
	// Can be used instead of `get_virtual_func`. Returns user data that will be passed to `call_virtual_func`.
	// Returning `NULL` from this function signals to Godot that the virtual function is not overridden.
	// Data returned from this function should be managed by the extension and must be valid until the extension is deinitialized.
	GDExtensionClassGetVirtualCallData get_virtual_call_data_func;
	// Used to call virtual functions when `get_virtual_call_data_func` is not null.
	GDExtensionClassCallVirtualWithData call_virtual_func;
```

fuzzybinary replied:
🤔 I don't think so? I'm using force pushes.... I didn't think I originally had the "Returns user data that will be passed to `call_virtual_func`" in the original documentation. But it also doesn't look as complete as I remember writing.

I'll take another stab at it.

Bromeon commented on core/extension/gdextension.cpp:
```diff
@@ -299,6 +299,8 @@ void GDExtension::_register_extension_class(GDExtensionClassLibraryPtr p_library
 		p_extension_funcs->free_instance_func, // GDExtensionClassFreeInstance free_instance_func; /* this one is mandatory */
 		p_extension_funcs->get_virtual_func, // GDExtensionClassGetVirtual get_virtual_func;
+		nullptr, // GDExtensionClassNotification2 get_virtual_call_data;
+		nullptr, // GDExtensionClassNotification2 call_virtual_func
```
These two types in the comment are not correct.

Nitpick, missing semicolon in 2nd line.

Bromeon reviewed (changes requested):
Few things to change, mostly clarify the semantics a bit and guide the user to choose between old vs. new approach.

dsnopek commented on core/extension/gdextension_interface.h:
```diff
@@ -265,9 +265,11 @@ typedef void (*GDExtensionClassToString)(GDExtensionClassInstancePtr p_instance,
+typedef GDExtensionObjectPtr (*GDExtensionClassCreateInstance)(void *p_class_userdata);
+typedef void (*GDExtensionClassFreeInstance)(void *p_class_userdata, GDExtensionClassInstancePtr p_instance);
+typedef GDExtensionClassCallVirtual (*GDExtensionClassGetVirtual)(void *p_class_userdata, GDExtensionConstStringNamePtr p_name);
+typedef void *(*GDExtensionClassGetVirtuaCallData)(void *p_class_userdata, GDExtensionConstStringNamePtr p_name);
```
This type-o which was pointed out by @Bromeon is still here:

```suggestion
typedef void *(*GDExtensionClassGetVirtualCallData)(void *p_class_userdata, GDExtensionConstStringNamePtr p_name);
```

I feel silly for having missed it in my review :-)

fuzzybinary replied:
🤦‍♂️ I'm the one who wrote it and missed it twice.

Should be fixed.

dsnopek reviewed (approved):
Looking good to me!

Bromeon reviewed (changes requested):
Two smaller things, see comments.

Side note, the names `get_virtual_call_data` and `call_virtual_with_data` are a bit difficult to memorize due to different order of the terms. Something like `get_virtual_data` and `call_virtual_data` might be a bit easier, although possibly less expressive. It's a tiny issue though and hardly relevant in practice, so if others are OK with the current naming, I won't bikeshed here 🙂

fuzzybinary commented:
> Two smaller things, see comments.
> 
> Side note, the names `get_virtual_call_data` and `call_virtual_with_data` are a bit difficult to memorize due to different order of the terms. Something like `get_virtual_data` and `call_virtual_data` might be a bit easier, although possibly less expressive.

Yeah, while easier, it's not actually descriptive of what's happening. Although harder to memorize with the reordering, I do think the parallelism of "get_call_data" -> "call_with_data" is helpful.

I'll rename the structure member to be more in line.

fuzzybinary commented:
@Bromeon Took one more stab at the documentation. Let me know if that's clearer.

Bromeon commented:
Mostly good, this is a tiny bit ambiguous due to unclear "precedence":

> You should supply either `get_virtual_func` or `get_virtual_call_data_func` with `call_virtual_with_data_func`.

Probably a comma would help a lot:

> You should supply either `get_virtual_func`, or `get_virtual_call_data_func` with `call_virtual_with_data_func`.

Other than that, good from my side, thanks for the adaptations!

fuzzybinary commented:
Comma added. Rebased on master.

Bromeon reviewed (approved):
Thanks a lot for your work! 👍

Bromeon commented:
There seems to be a memory corruption in the `double` build, [here](https://github.com/godotengine/godot/actions/runs/6243304110/job/16948531395?pr=80671#step:18:224)... Not sure if introduced by this PR?

akien-mga commented:
Thanks!
